armory/art_experimental/attacks/snr_pgd.py

```diff
@@ -1,11 +1,235 @@
-from art.attacks.evasion import ProjectedGradientDescentPyTorch
+from art.attacks.evasion import (
+    ProjectedGradientDescentPyTorch,
+    ProjectedGradientDescentNumpy,
+)
 import numpy as np
 
 import logging
 
 logger = logging.getLogger(__name__)
 
 
+class SNR_PGDRange:
+    """
+    Finds the maximum SNR for each sample (if possible), via binary search.
+
+    Example attack config:
+        "attack": {
+            "kwargs": {
+                "batch_size": 1,
+                "eps_range": [
+                    10,
+                    50
+                ],
+                "eps_step": 0.5,
+                "max_iter": 10,
+                "norm": "snr_db",
+                "num_random_init": 2,
+                "targeted": false,
+                "verbose": false
+            },
+            "module": "armory.art_experimental.attacks.snr_pgd",
+            "name": "SNR_PGDRange",
+            "use_label": true
+        }
+    """
+
+    def __init__(self, estimator, eps_range=(0, 10, 20, 30, 40, 50), **kwargs):
+        if "eps" in kwargs:
+            raise ValueError("Use 'eps_range' instead of 'eps'")
+        self.estimator = estimator
+        self.eps_range = sorted(eps_range)
+        if len(eps_range) < 2:
+            raise ValueError("Please select multiple values for eps_range")
+        self.attacks = []
+        for eps in eps_range:
+            self.attacks.append(SNR_PGD(estimator, eps=eps, **kwargs))
+
+    def generate(self, x, y=None, **kwargs):
+        if y is None:
```

Comment:
Me may want to assert that `kwargs["targeted"] == False` both here and in the `SNR_PGDRange2` variant. Otherwise, the logic of detecting attack success/failure by comparing to `y` will be backwards.